### PR TITLE
Selenium session teardown

### PR DESCRIPTION
**Pull request by sawantpritam**
@itsankit-google @rmstar  Closing selenium driver session after all scenarios. 
Cucumber does not have @BeforeAll and @AfterAll tags. This is workaround using Runtime.getRuntime().addShutdownHook in @Before.

### CODE REVIEW

File: src/main/java/stepsdesign/AfterActions.java [hunk 1]
```diff
@@ -30,9 +30,13 @@
 public class AfterActions {
 
   @AfterStep
-  public static void tearDown(Scenario scenario) {
-    WebDriver driver = SeleniumDriver.getDriver();
-    byte[] screenshotBytes = ((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES);
-    scenario.embed(screenshotBytes, "image/png");
+  public static void takeScreenshot(Scenario scenario) {
```
Comment (rmstar) [inline src/main/java/stepsdesign/AfterActions.java:33]: where is takeScreenShot called?

Comment (sawantpritam) [reply src/main/java/stepsdesign/AfterActions.java:33]: It is not called explicitly anywhere. It gets executed after every cucumber test step with`@AfterStep` hook

File: src/main/java/stepsdesign/BeforeActions.java [hunk 2]
```diff
@@ -32,11 +32,16 @@
 public class BeforeActions {
   public static Scenario scenario;
   public static File myObj;
+  public static boolean beforeAllFlag = true;
 
   @Before(order = 0)
   public void setUp(Scenario scenario) throws IOException {
+    if (beforeAllFlag) {
```
Comment (rmstar) [inline src/main/java/stepsdesign/BeforeActions.java:39]: why do we need this flag? Is setUp called multiple times?

Comment (rmstar) [reply src/main/java/stepsdesign/BeforeActions.java:39]: does it need to be volatile?

**Comment by rmstar** [comment]
could we do something like @AfterStep(order=1000) to ensure that the cleanup step is executed at the end?

File: src/main/java/stepsdesign/BeforeActions.java (hunk 2 above)
Comment (sawantpritam) [reply src/main/java/stepsdesign/BeforeActions.java:39]: Yes Cucumber `@Before` `setUp()` gets executed before each scenario.
Cucumber does not have `@BeforeAll `and `@Afterall` tags which will execute before first and after last scenario. We want to setup selenium driver before first scenario and teardown after final scenario. So flag is needed.

**Comment by sawantpritam** [comment]
> could we do something like @afterstep(order=1000) to ensure that the cleanup step is executed at the end?

Order for @AfterStep is not required. The default value is 10000 if not mentioned explicitly.

**Review by rmstar — APPROVED** [review]

**Comment by itsankit-google** [comment]
> could we do something like @afterstep(order=1000) to ensure that the cleanup step is executed at the end?

Its actually something good in cucumber that afterstep order is reverse, for eg, order=0 in after will be executed at the end.

**Pull request merged**